index.ts

```diff
@@ -315,6 +315,233 @@ export async function createElasticsearchMcpServer(
     }
   );
 
+  // Tool 4: Get shard information
+  server.tool(
+    "get_shards",
+    "Get shard information for all or specific indices",
+    {
+      index: z
+        .string()
+        .optional()
+        .describe("Optional index name to get shard information for"),
+    },
+    async ({ index }) => {
+      try {
+        const response = await esClient.cat.shards({
+          index,
+          format: "json",
+        });
+
+        const shardsInfo = response.map((shard) => ({
+          index: shard.index,
+          shard: shard.shard,
+          prirep: shard.prirep,
+          state: shard.state,
+          docs: shard.docs,
+          store: shard.store,
+          ip: shard.ip,
+          node: shard.node,
+        }));
+
+        const metadataFragment = {
+          type: "text" as const,
+          text: `Found ${shardsInfo.length} shards${index ? ` for index ${index}` : ""}`,
+        };
+
+        return {
+          content: [
+            metadataFragment,
+            {
+              type: "text" as const,
+              text: JSON.stringify(shardsInfo, null, 2),
+            },
+          ],
+        };
+      } catch (error) {
+        console.error(
+          `Failed to get shard information: ${
+            error instanceof Error ? error.message : String(error)
+          }`
+        );
+        return {
+          content: [
+            {
+              type: "text" as const,
+              text: `Error: ${
+                error instanceof Error ? error.message : String(error)
+              }`,
+            },
+          ],
+        };
+      }
+    }
+  );
+
+  // Tool 5: Query Profiler
+  server.tool(
+    "profile_query",
+    "Analyze query performance using Elasticsearch Profile API",
+    {
+      index: z
+        .string()
+        .trim()
+        .min(1, "Index name is required")
+        .describe("Name of the Elasticsearch index to profile"),
+
+      queryBody: z
+        .record(z.any())
+        .refine(
+          (val) => {
+            try {
+              JSON.parse(JSON.stringify(val));
+              return true;
+            } catch (e) {
+              return false;
+            }
+          },
+          {
+            message: "queryBody must be a valid Elasticsearch query DSL object",
+          }
+        )
+        .describe("Elasticsearch query DSL to profile"),
+
+      explain: z
+        .boolean()
+        .optional()
+        .default(false)
+        .describe("Whether to include explanation of how the query was executed"),
+    },
+    async ({ index, queryBody, explain }) => {
+      try {
+        const searchRequest = {
+          index,
+          body: {
+            ...queryBody,
+            profile: true,
+            explain: explain,
```

Comment:
Given that profile/explain is part of search API, isn't it sort of supported by `search` tool?

Would it make sense to just add new param to search tool e.g. `explain: bool` that would:
- set profile and explain to true
- process the response and include profile info in text fragmetns

Reply:
[7fe9331](https://github.com/elastic/mcp-server-elasticsearch/pull/28/commits/7fe9331797dd5ec61a6ed2190300e4c7a850b251)

Previously, we had a separate profile_query tool for query profiling, but I've now integrated these features into the search tool. This change adds profile and explain parameters to the existing search functionality, allowing users to get profiling data and query explanations in a single request. The integration includes performance analysis and optimization suggestions